### PR TITLE
Add help slash command

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/riot-web/issues/10289

image

Signed-off-by: Michael Telatynski <7t3chguy@gmail.com>

### CODE REVIEW

**Comment by t3chguy** [comment]
image

File: src/components/views/dialogs/InfoDialog.js [hunk 1]
```diff
@@ -28,12 +28,14 @@ export default React.createClass({
         description: PropTypes.node,
         button: PropTypes.string,
         onFinished: PropTypes.func,
+        hasCloseX: PropTypes.bool,
```
Comment (turt2live) [inline src/components/views/dialogs/InfoDialog.js]: tbh I think I'd prefer `hasCloseButton`

Comment (dbkr) [reply src/components/views/dialogs/InfoDialog.js]: I think `hasCancel` would match what we have elsewhere, for better or worse.

Comment (t3chguy) [reply src/components/views/dialogs/InfoDialog.js]: That was my original wording but then it left ambiguity as two components within InfoDialog have a `hasCancel` prop, the dialog itself (for the closing X) and DialogButtons, so it'd be unclear which/both it affected

Comment (turt2live) [reply src/components/views/dialogs/InfoDialog.js]: *everything is awful themesong plays*

as much as I normally argue for consistency, `hasCloseButton` is probably safer given @t3chguy's comment imo.

**Comment by nadonomy** [comment]
Thanks for this! Small questions/potential polish requests:

- What happens on smaller screens/if the modal body content doesn't fit on the screen? (Presumably it becomes scrollable?)
- Non-developers won't be familiar with the nomenclature 'Arguments' and I don't think it's something we want to teach, so I'd be in favour of just dropping the title row altogether as the columns will still make sense contextually
- Can we add some padding between the columns, and increase the font weight of the commands column to provide better contrast and legibility when scanning?
- Is it trivial to split the table up into sections, like the Preferences tab of Settings? If so, let me know and I can quickly group & name them into sections that make sense.

**Comment by t3chguy** [comment]
> What happens on smaller screens/if the modal body content doesn't fit on the screen? (Presumably it becomes scrollable?)

 image

> Non-developers won't be familiar with the nomenclature 'Arguments' and I don't think it's something we want to teach, so I'd be in favour of just dropping the title row altogether as the columns will still make sense contextually

Sure

> Can we add some padding between the columns, and increase the font weight of the commands column to provide better contrast and legibility when scanning?

Sure

> Is it trivial to split the table up into sections, like the Preferences tab of Settings? If so, let me know and I can quickly group & name them into sections that make sense.

Sure

--
will apply changes 2&3, 4 once I get the groupings from you :)

**Comment by nadonomy** [comment]
@t3chguy thanks! I just put together some grouping here which I think makes sense:

https://docs.google.com/document/d/14Pjtkue-2NfM2qlXBLG0GVhfn4lwYfO_S-k6Oi4f9Tw/edit?usp=sharing

Feedback/a second pair of eyes is welcome, let me know if you think anything looks off.

**Comment by t3chguy** [comment]
I might have to mangle this into a single table to make the width consistent:

image

**Comment by turt2live** [comment]
colspan ftw

**Comment by t3chguy** [comment]
> colspan ftw

I know, but it still feels dirty :P

**Comment by t3chguy** [comment]
image

File: src/SlashCommands.js [hunk 2]
```diff
@@ -274,30 +274,31 @@ export const CommandMap = {
...
-                    }
-                    return success(
-                        SettingsStore.setValue('roomColor', roomId, SettingLevel.ROOM_ACCOUNT, colorScheme),
-                    );
-                }
-            }
-            return reject(this.getUsage());
-        },
-    }),
+    // Room Tint has been unsupported since the redesign
```
Comment (turt2live) [inline src/SlashCommands.js]: I'd just delete the command for now. We have git history if we need it.

File: res/css/views/dialogs/_SlashCommandHelpDialog.scss [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/*!
```
Comment (turt2live) [inline res/css/views/dialogs/_SlashCommandHelpDialog.scss]: You've got some weird copyright going on, which seems duplicated below.

**Comment by turt2live** [comment]
... assuming the CI is happy

File: res/css/views/dialogs/_SlashCommandHelpDialog.scss (hunk 3 above)
Comment (t3chguy) [reply res/css/views/dialogs/_SlashCommandHelpDialog.scss]: Arg this is WebStorm thinking it is helping but it infact is not :(

File: src/SlashCommands.js (hunk 2 above)
Comment (t3chguy) [reply src/SlashCommands.js]: Sure

**Comment by Biep** [comment]
I like it!
Any chance for arguments to /help for more detailed explanation where appropriate?  Many casual users may not know what a room alias is, for instance. so _/help join_ might produce a less terse and more beginner-friendly explanation.

**Comment by turt2live** [comment]
@Biep that's better suited for a new issue

**Comment by Biep** [comment]
OK - but it might help in implementing the general */help* command to know that a form with arguments might be wanted too at some point.
By default the form with argument would just reproduce the relevant line (or section, if the argument is something like *Messages*) from the main */help* output.  Only if specific help is available it would show that.

**Pull request merged**